dispatcher/nm-dispatcher.c

```diff
@@ -575,29 +575,38 @@ dispatch_one_script (Request *request)
 	return FALSE;
 }
 
+static int
+_compare_basenames (gconstpointer a, gconstpointer b)
+{
+	const char *basename_a = strrchr (a, '/');
+	const char *basename_b = strrchr (b, '/');
+	int ret;
+
+	g_return_val_if_fail (basename_a, 0);
+	g_return_val_if_fail (basename_b, 0);
+
+	ret = strcmp (++basename_a, ++basename_b);
+	if (ret)
+		return ret;
+
+	return strcmp (a, b);
```

Comment:
ok, `basename_a` and `basename_b` are the same.

That means, `/etc/NetworkManager/.../$basename_a` shadows `/usr/lib/NetworkManager/.../$basename_a`. Just comparing them with `strcmp()` does not correctly prefer `/etc` or `/usr`. Ok, the final final `strcmp()` doesn't really matter here, because there are larger issues. You not only need to sort such two entries, you need to skip the shadowed one from running! But preferably, not without a debug message in the log.


Also, `_find_scripts()` only return scripts that are executable. Wouldn't it be nice, if the user could prevent `/usr/lib/NetworkManager/dispatcher.d/$SCRIPT` from running, by putting a non-executable file in `/etc` (or a symlink to `/dev/null`)? That means, `_find_scripts()` must return all paths in the first moment, and collect the information about whether it's executable or not, and use that to figure out which scripts are shadowed.

Reply:
maybe this is the point, where we need unit-tests for this code...

Reply:
I think you not only track the filenames, but also (at least) `struct stat` for the permission check. If you happen to do that, I think we should not use `GSList`. That is just the worst data structure.